### PR TITLE
feat: [metrics] add support for float

### PR DESCRIPTION
**Pull request by lalo**

### CODE REVIEW

**Review by jackgerrits — APPROVED** [review]

File: vowpalwabbit/metrics.cc [hunk 1]
```diff
@@ -44,10 +44,15 @@ void list_to_json_file(std::string filename, std::vector<std::tuple<std::string,
...
 
     writer.StartObject();
-    for (std::tuple<std::string, size_t> m : metrics)
+    for (std::tuple<std::string, int32_t> m : metrics.int_metrics_list)
     {
       writer.Key(std::get<0>(m).c_str());
-      writer.Int64(static_cast<int32_t>(std::get<1>(m)));
+      writer.Uint64(std::get<1>(m));
+    }
+    for (std::tuple<std::string, int32_t> m : metrics.float_metrics_list)
```
Comment (lalo) [inline vowpalwabbit/metrics.cc:52]:
```suggestion
    for (std::tuple<std::string, float> m : metrics.float_metrics_list)
```

File: vowpalwabbit/metrics.cc [hunk 2]
```diff
@@ -44,10 +44,15 @@ void list_to_json_file(std::string filename, std::vector<std::tuple<std::string,
     Writer<FileWriteStream> writer(os);
 
     writer.StartObject();
-    for (std::tuple<std::string, size_t> m : metrics)
+    for (std::tuple<std::string, size_t> m : metrics.int_metrics_list)
     {
       writer.Key(std::get<0>(m).c_str());
```
Comment (jackgerrits) [inline vowpalwabbit/metrics.cc:49]:
```suggestion
      writer.Key(m->first.c_str());
```

Comment (jackgerrits) [reply vowpalwabbit/metrics.cc:49]: I prefer using the iterator and then pair accessors

File: vowpalwabbit/cb_explore_adf_common.h [hunk 3]
```diff
@@ -90,7 +92,7 @@ struct cb_explore_adf_base
   static void print_multiline_example(vw& all, cb_explore_adf_base<ExploreType>& data, multi_ex& ec_seq);
   static void save_load(cb_explore_adf_base<ExploreType>& data, io_buf& io, bool read, bool text);
   static void persist_metrics(
-      cb_explore_adf_base<ExploreType>& data, std::vector<std::tuple<std::string, size_t>>& metrics);
+      cb_explore_adf_base<ExploreType>& data, std::vector<std::tuple<std::string, int32_t>>& metrics);
```
[the remark this answers is not in the document]
Comment (lalo) [reply vowpalwabbit/cb_explore_adf_common.h]: not anymore

File: vowpalwabbit/metric_sink.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+namespace VW
+{
+struct metric_sink
+{
+  std::vector<std::tuple<std::string, size_t>> int_metrics_list;
```
[the remark this answers is not in the document]
Comment (lalo) [reply vowpalwabbit/metric_sink.h]: it can be unsigned

**Review by olgavrou — APPROVED** [review]

**Pull request merged**